controllers/k8ssandra/medusa_reconciler.go

```diff
@@ -0,0 +1,385 @@
+	kc *api.K8ssandraCluster,
+	logger logr.Logger,
+) (ctrl.Result, error) {
+	// Medusa isn't enabled (anymore) in this cluster
+	logger.Info("Medusa reconcile for " + dcConfig.Meta.Name + " on namespace " + kc.Namespace + "/" + dcConfig.Meta.Namespace)
+	medusaSpec := kc.Spec.Medusa
+	if medusaSpec != nil {
+		logger.Info("Medusa is enabled")
+		if dcConfig.PodTemplateSpec == nil {
+			dcConfig.PodTemplateSpec = &corev1.PodTemplateSpec{
+				Spec: corev1.PodSpec{
+					Containers:     []corev1.Container{},
+					InitContainers: []corev1.Container{},
+				},
+			}
+		}
+		if result, err := r.reconcileMedusaConfigMap(ctx, remoteClient, kc, logger); !result.IsZero() || err != nil {
+			return result, err
+		}
+		updateMedusaInitContainer(dcConfig, medusaSpec, logger)
+		updateMedusaMainContainer(dcConfig, medusaSpec, logger)
+		updateMedusaVolumes(dcConfig, medusaSpec, logger)
+	} else {
+		logger.Info("Medusa is not enabled")
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// Generate a secret for Medusa or use the existing one if provided in the spec
+func (r *K8ssandraClusterReconciler) reconcileMedusaSecrets(
+	ctx context.Context,
+	kc *api.K8ssandraCluster,
+	logger logr.Logger,
+) error {
+	logger.Info("Reconciling Medusa user secrets")
+	if kc.Spec.Medusa != nil {
+		cassandraUserSecretRef := kc.Spec.Medusa.CassandraUserSecretRef
+		if cassandraUserSecretRef == "" {
+			cassandraUserSecretRef = cassandraUserSecretName(kc.Spec.Medusa, kc.Name)
+		}
+		logger = logger.WithValues(
+			"MedusaCassandraUserSecretRef",
+			cassandraUserSecretRef,
+		)
+		if err := secret.ReconcileSecret(ctx, r.Client, cassandraUserSecretRef, kc.Name, kc.Namespace); err != nil {
+			logger.Error(err, "Failed to reconcile Medusa CQL user secret")
+			return err
+		}
+	}
+	logger.Info("Medusa user secrets successfully reconciled")
+	return nil
+}
+
+// Create the Medusa config map if it doesn't exist
+func (r *K8ssandraClusterReconciler) reconcileMedusaConfigMap(
+	ctx context.Context,
+	remoteClient client.Client,
+	kc *api.K8ssandraCluster,
+	logger logr.Logger,
+) (ctrl.Result, error) {
+	logger.Info("Reconciling Medusa configMap on namespace : " + kc.Namespace)
+	if kc.Spec.Medusa != nil {
+		medusaIni := createMedusaIni(kc)
+		desiredConfigMap := createMedusaConfigMap(kc.Spec.Medusa, kc.Namespace, kc.Spec.Cassandra.Cluster, medusaIni)
+		// Compute a hash which will allow to compare desired and actual configMaps
+		utils.AddHashAnnotation(desiredConfigMap, api.ResourceHashAnnotation)
+		actualConfigMap := &corev1.ConfigMap{}
+
+		configMapKey := client.ObjectKey{
+			Namespace: kc.Namespace,
+			Name:      fmt.Sprintf("%s-medusa", kc.Spec.Cassandra.Cluster),
+		}
+
+		if err := remoteClient.Get(ctx, configMapKey, actualConfigMap); err != nil {
+			if errors.IsNotFound(err) {
+				if err := remoteClient.Create(ctx, desiredConfigMap); err != nil {
+					logger.Error(err, "Failed to create Medusa ConfigMap")
+					return ctrl.Result{RequeueAfter: r.DefaultDelay}, err
```

Comment:
You don't need to set `RequeueAfter` when the error is not nil. The request will requeued with an exponential backoff by default that is configurable via controller-runtime.

Reply:
This was changed since the last rebase as reconcile result handling was wrapped the ReconcileResult type.